### DISCUSSION TITLE
Playlist.saveWithUpdate дублирует треки при некоторых обстоятельствах

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#199] [Помощь с ошибкой]
Есть серия плейлистов, в которые я добавляю треки вручную и мне важно чтобы в нём оставалась дата добавления трека. Триггер удаляет треки попадающие в бан. 
Все плейлисты собраны по такому типу и размещены в одном файле с разными именами: 
```
  let bannedTracks = Cache.read('ban.json');

  let tracks = Source.getPlaylistTracks('плейлист', 'id');
  Filter.removeTracks(tracks, bannedTracks);

  Playlist.saveWithUpdate({
    id: 'id',
    name: 'плейлист ( ' + tracks.length + ' )',
    tracks: tracks,
  });
```
Всё работало как часы. Несколько месяцев назад была большая разовая серия ошибок, когда в отчётах было указано, что многим триггерам не хватало 6 минут на выполнение (превышено максимальное время на выполнение).
После этого, не уверен, что именно вследствие этого, количество треков в плейлистах резко увеличилось (появились дубликаты). Некоторые треки были продублированы несколько раз. Я решил, что это какойто-то глобальный сбой в приложении Apps Script, и пожертвовав датами добавления треков удалил дубликаты с помощью `Filter.dedupTracks` и `Playlist.saveWithReplace`. После этого всё работало без нареканий. 
Вчера количество треков в одном из плейлистов снова увеличилось примерно в два раза за счёт появления дубликатов. Посмотрел ошибки выполнения триггера...
![2022-08-07_15-52-38](https://user-images.githubusercontent.com/85184404/183287553-e997bcad-047f-4cea-bbea-66d851a61cab.png)
Служба была вызвана слишком много раз за день. И так все 6 минут... Не понимаю почему. Вроде всё как всегда...

Собственно в этом проблема. Может есть способ избежать такого массового дубляжа при ошибках выполнения?
Дубликаты удалю вручную на десктопе, благо по дате видно гигантское последнее прибавление. 

Кстати, сейчас только обратил внимание, что продублировались все треки начная со 101 по последний. Первые сто треков не продублировались...

**Comment by Chimildic** [comment]
Попробуйте посмотреть логи других триггеров. Возможно кто-то нарвался на ошибку и в попытках повторной отправки потратил весь лимит.

**Reply by Bondian** [reply]
Уже даже не сомневаюсь, что лимит потрачен или другими триггерами или моими многочисленными экспериментами с аддоном FM-радио... Но вопрос несколько шире...
> ... Может есть способ избежать такого массового дубляжа при ошибках выполнения?

Имеется ввиду выполнение метода сохранения плейлиста `Playlist.saveWithUpdate`

**Reply by Chimildic** [reply]
Довольно загадочная ситуация. Судя по логу он прерывается при получении треков плейлиста. После нескольких попыток неудачных запросов пойдет дальше с пустым массивом. То есть фильтр ничего не удалит. Но почему ему потом позволяют отправить запросы при исчерпанном лимите, чтобы сложилась ситуация с дублями?.. Или мы просто не туда смотрим.

Попробуйте целиком заменить функцию goofy `function tryCallback(callback, attempt = 0)` на следующий код:
```js
function tryCallback(callback, attempt = 0) {
    try {
        return callback();
    } catch (error) {
        Admin.printError('При отправке запроса произошла ошибка:\n', error.stack);
        if (error.stack.toLowerCase().includes("urlfetch.")) {
          throw "Выполнение прервано принудительно"
        } else if (attempt++ < 2) {
            Admin.pause(5);
            return tryCallback(callback, attempt);
        }
    }
}
```
Если в ошибке встретится `urlfetch`, скрипт прервется. Поэтому дальше по коду он просто не пойдет.

К слову, эксперименты с радио действительно могут исчерпать лимит. Потому что каждый трек это один запрос для поиска. И альтернатив этому нет. Попробуйте ставить лимит на количество треков во время эксперимета. У функций `getTracks` и `getTopTracks` он идет вторым аргументом.

**Reply by Bondian** [reply]
Функцию заменил, но ошибок пока нет. Надо признать - они редки... Буду наблюдать.
Если библиотека обновится - надо будет снова заменить функцию?

**Reply by Chimildic** [reply]
Включу в основу, если сработает как предполагаем

**Reply by Bondian** [reply]
Пропустил много почтовых уведомлений. Оказывается ошибки были почти каждый день в одно и то же время, хотя триггер запускался каждые четыре часа. Но дублирование треков произошло только один раз за последнее время. Раньше, после вышеуказанных скриншотов с повторами, они заканчивались вот так (не заметил ранее): 
![концовка до](https://user-images.githubusercontent.com/85184404/183564586-b23bde46-c9d0-4ed7-9034-bc7a3ee29d24.png)
Теперь свежая ошибка, Опять в первый час суток, но теперь, после замены функции, выполнение сразу прервалось, как и было необходимо.
![после](https://user-images.githubusercontent.com/85184404/183565097-99ff11e4-2204-48d6-b383-2e527e88cda2.png)

**Reply by Chimildic** [reply]
Похоже, что у вас настроено слишком много задач и они ежедневно превышают лимит.
Посмотрите статистику в [консоли Spotify](https://developer.spotify.com/dashboard). Сколько запросов у _Total Number of Requests_?

**Reply by Bondian** [reply]
> Похоже, что у вас настроено слишком много задач и они ежедневно превышают лимит.

Так и есть. Буду принимать меры по экономии запросов.
![2022-08-09_11-52-22](https://user-images.githubusercontent.com/85184404/183584941-4cf5bd1b-d58b-4cda-8350-0f6c0ad2a447.png)
